jwt.js

```diff
@@ -186,42 +188,66 @@ function fastifyJwt (fastify, options, next) {
     return token
   }
 
-  function sign (payload, options, callback) {
-    assert(payload, 'missing payload')
+  function mergeOptionsWithKey (options, useProvidedPrivateKey) {
+    if (useProvidedPrivateKey && (typeof useProvidedPrivateKey !== 'boolean')) {
+      return Object.assign({}, options, { key: useProvidedPrivateKey })
+    } else {
+      const key = useProvidedPrivateKey ? secretOrPrivateKey : secretOrPublicKey
+      return Object.assign(!options.key ? { key } : {}, options)
+    }
+  }
+
+  function checkAndMergeOptions (options, defaultOptions, usePrivateKey, callback) {
+    let mergedOptions
 
     if (typeof options === 'function') {
       callback = options
-      options = Object.assign({}, signOptions)
+      mergedOptions = mergeOptionsWithKey(defaultOptions, usePrivateKey)
+    } else {
+      if (!options) {
+        mergedOptions = mergeOptionsWithKey(defaultOptions, usePrivateKey)
+      } else {
+        mergedOptions = mergeOptionsWithKey(options, usePrivateKey)
+      }
     }
 
-    if (!options) {
-      options = Object.assign({}, signOptions)
-    }
+    return { options: mergedOptions, callback }
+  }
+
+  function checkAndMergeSignOptions (options, callback) {
+    return checkAndMergeOptions(options, signOptions, true, callback)
+  }
+
+  function checkAndMergeVerifyOptions (options, callback) {
+    return checkAndMergeOptions(options, verifyOptions, false, callback)
+  }
+
+  function sign (payload, options, callback) {
+    assert(payload, 'missing payload')
 
-    if (typeof callback === 'function') {
-      jwt.sign(payload, secretOrPrivateKey, options, callback)
+    const signerConfig = checkAndMergeSignOptions(options, callback)
+    const signer = createSigner(signerConfig.options)
```

Comment:
We should not be creating a signer every time. It's expensive and remove most of the benefits of fast-jwt.